src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs

```diff
@@ -94,22 +108,93 @@ internal BlameCollector(IBlameReaderWriter blameReaderWriter, IProcessDumpUtilit
+
+            if (this.collectProcessDumpOnTestHostHang)
+            {
+                this.inactivityTimer = new InactivityTimer(this.CollectDumpAndAbortTesthost);
+                this.ResetInactivityTimer();
+            }
+        }
+
+        /// <summary>
+        /// Disposes of the timer when called to prevent further calls.
+        /// Kills the other instance of procdump if launched for collecting trigger based dumps.
+        /// Starts and waits for a new proc dump process to collect a single dump and then
+        /// kills the testhost process.
+        /// </summary>
+        private void CollectDumpAndAbortTesthost()
+        {
+            EqtTrace.Info(string.Format(CultureInfo.CurrentUICulture, Resources.Resources.InactivityTimeout, (int)this.inactivityTimespan.TotalMinutes));
+            this.inactivityTimerAlreadyFired = true;
+
+            try
+            {
+                EqtTrace.Verbose("Calling dispose on Inactivity timer.");
+                this.inactivityTimer.Dispose();
+            }
+            catch
+            {
+                EqtTrace.Verbose("Inactivity timer is already disposed.");
+            }
+
+            if (this.collectProcessDumpCollectionBasedOnTrigger)
+            {
+                // Detach procdump from the testhost process to prevent testhost process from crashing
+                // if/when we try to kill the existing proc dump process.
+                this.processDumpUtility.DetachFromTargetProcess();
+            }
+
+            try
+            {
+                this.processDumpUtility.StartHangBasedProcessDump(this.testHostProcessId, this.attachmentGuid, this.GetResultsDirectory(), this.processFullDumpEnabled);
+            }
+            catch (Exception ex)
+            {
+                EqtTrace.Error($"BlameCollector.CollectDumpAndAbortTesthost: Failed with error {ex}");
+            }
+
+            try
+            {
+                var dumpFile = this.processDumpUtility.GetDumpFile();
+                if (!string.IsNullOrEmpty(dumpFile))
+                {
+                    var fileTransferInformation = new FileTransferInformation(this.context.SessionDataCollectionContext, dumpFile, true);
+                    this.dataCollectionSink.SendFileAsync(fileTransferInformation);
+                }
+                else
+                {
+                    EqtTrace.Error("BlameCollector.CollectDumpAndAbortTesthost: blame:CollectDumpOnHang was enabled but dump file was not generated.");
+                }
+            }
+            catch (FileNotFoundException ex)
+            {
+                EqtTrace.Error(ex);
+            }
+
+            try
+            {
+                Process.GetProcessById(this.testHostProcessId).Kill();
+            }
+            catch (Exception ex)
+            {
+                EqtTrace.Error(ex);
+            }
         }
 
         private void ValidateAndAddProcessDumpParameters(XmlElement collectDumpNode)
```

Comment:
Lets refactor this function, before this code goes in, it won't be possible in future.